README.md
# ChessBrainVB
Chess engine with GUI for Excel / Word / Powerpoint VBA - plus edition for UCI/Winboard+SMP: win32 engine with playing strength of 3150 ELO (4 CPU)

There are two ways to use this chess engine:

1. Use a free chess GUI like ARENA, add ChessBrainVB.exe as UCI or winboard engine  and play games.
   Files needed: ChessBrainVB.ini  for settings, ChessBrainVB_Book.opn, ChessBrainVB_Book.opi for opening book.
  Playing strength 3150 ELO (CCRL 40/40 conditions, 4CPU, see http://www.computerchess.org.uk/ccrl/4040/rating_list_all.html)
  Compiled with Visual Basic 6
  Since V3.65: Multi core version for up to 64 threads, maximum hash size 1.4 GB.
 
2. Use ExcelChessBrainX.xlsm, WordChessBrainX.docm or PowerpointChessBrainX.pptm (full install needed, viewer not working)
   to play games using the GUI implemented in VBA forms.
  Files needed: ChessBrainVB.ini  for settings, ChessBrainVB_Book.opn, ChessBrainVB_Book.opi for opening book.
  Playing strength 2500 ELO
  Visual Basic for Applications is used for the chess engine.
  This pseudo code is not compiled and 15 times slower than the compiled VB6 EXE for winboard.

Hint: The program "Alice in Chessland" by Angelo Spartalis has a very nice GUI and is based on ChessbrainVB V3.05. 
Link: https://www.spartalis.gr/chess/index_ENG.html

For questions please contact me:
rogzuehlsdorf@yahoo.de
----------------------------------------------------------------------
### CREDITS
This chess engine is based on the source of the engine "LarsenVB" by Luca Dormio (http://xoomer.virgilio.it/ludormio/download.htm).
LarsenVB was inspired by "Faile 0.6 by" Adrien M. Regimbald, which was also the base for the engine "Sjeng".
I want to thank Luca Dormio for his permission to use his LarsenVB source. 

ChessBrainVB is also based on many great ideas from the following people: 

Marco Costabla/Tord Romstad/Joona Kiiski (Stockfish sources): Search logic, king safety, piece evaluation.
Search logic and evaluation are based an Stockfish 7 with adaptions to non-bitboard data structure and search changes that perform better for slower move generation and evaluation.
Raimund Heid (Protector sources):  Material draw logic
Norbert Raimund Leisner: Logo file

----------------------------------------------------------------------
Keywords: "Excel chess engine", "Word chess engine", "Powerpoint chess engine", "VBA chess engine", "VB6 chess engine", "VBA chess game", "Excel chess game", "Visual Basic chess program"


Release/ReleaseNotes.txt
=======================================================================================
= ReleaseNotes.txt
=
= Release notes for ChessBrainVB chess engine 
=======================================================================================

Changes from version 3.72 to 3.74 (Roger Zuehlsdorf 09-AUG-2019)
----------------------------------------------------------------
- minor fixes
- code cleanup
- ELO gain: about 10 ELO 

Changes from version 3.70 to 3.72 (Roger Zuehlsdorf 02-DEC-2018)
----------------------------------------------------------------
- less optimistic king safety evaluation
- 10% faster: slow VB6 assignments for user defined data types replaced
- code cleanup
- ELO gain: about 30 ELO for short time controls

Changes from version 3.68 to 3.70 (Roger Zuehlsdorf 11-JUL-2018)
----------------------------------------------------------------
- bug fixes in evaluation
- code cleanup
- ELO gain: about 20 ELO

Changes from version 3.67 to 3.68 (Roger Zuehlsdorf 1-MAR-2018)
----------------------------------------------------------------
- fixed bug: fifty move counter not seen by helper threads 
- code cleanup

Changes from version 3.61 to 3.65 (Roger Zuehlsdorf 24-FEB-2018)
----------------------------------------------------------------
- fixed bug: enpassant capture ignored in mate escape move list (reason for game loosing moves)
  this old bug was discovered in engine matches (many thanks to Graham Banks for running this matches)
- some endgame knowledge added for positions with no pawns

Changes from version 3.61 to 3.65 (Roger Zuehlsdorf 18-FEB-2018)
----------------------------------------------------------------
- new time management, moveoverhead setting added
- UCI fixes, hashfull info added, Chessbase GUI supported using InBetween.exe
- hash file limit 1400MB for compiled exe (IDE and VBA still 384)
- SMP improvements: 
    support for 64 cores
	helper processess starting much faster
	verify hash read/write option to detect hash collisions
- ELO gain on self play: 40/4 about 20 ELO from time management, no improvement for 40/40  

Changes from version 3.60 to 3.61 (Roger Zuehlsdorf 25-DEC-2017)
----------------------------------------------------------------
- UCI time management problems fixed (tested with ARENA, other GUIs should use Winboard interface)
- UCI buffer overflow problem fixed with long matches (more than 200 moves)
- time management optimized for sudden death games


Changes from version 3.50 to 3.60 (Roger Zuehlsdorf 17-DEC-2017)
----------------------------------------------------------------
this is a feature release:
- support for UCI protocol (includes settings for number of cores and endgame tablebases)
- very limited support for Syzygy endgame tablebases using the fathom.exe tool output
  (limited to first the first 3 plies because slow - about 10 positions per second)
- code cleanup and changes in singular extensions
- ELO gain in self-play against V3.50 is small: 40/4 about 10 ELO, may be more for 40/40

Changes from version 3.31 to 3.50 (Roger Zuehlsdorf 26-AUG-2017)
----------------------------------------------------------------
- multi-processor support up to 8 threads (not for VBA office)
  using CreateFileMapping-API for shared hash and PV. (see file MultiCore_Notes.txt)
  Note: Used memory may not show up in task mananger (but displayed correctly in ARENA)
- keep history values for next game move (don't delete values at begin of search)
- better move count pruning exceptions for threat move evasions
- bug fix in pawn eval
- speed optimized
- ELO gain in self-play against V3.31: 60 ELO at 40/4, 30 ELo at 40/40.

Changes from version 3.20 to 3.31 (Roger Zuehlsdorf 4-JUN-2017)
---------------------------------------------------------------
- several small fixes and code cleanup 
- speed optimized
- changed move heuristic
- ELO increase about 20 ELO
- time management bug fixed

Changes from version 3.03 to 3.20 (Roger Zuehlsdorf 10-MAR-2017)
----------------------------------------------------------------
- bugfix in king safety
- several small fixes and code cleanup 
- move counter added to Office GUI
- ELO increase about 20 ELO

Changes from version 3.03 to 3.03a (Roger Zuehlsdorf 15-DEC-2016)
----------------------------------------------------------------
- compact edition for WORD and POWERPOINT added


Changes from version 3.02 to 3.03 (Roger Zuehlsdorf 13-DEC-2016)
----------------------------------------------------------------
- VBA overflow at depth 100 fixed
- VBA move list output fixed


Changes from version 2.05 to 3.02 (Roger Zuehlsdorf 09-DEC-2016)
----------------------------------------------------------------
- major rewrite of move generation, king safety and search logic (new stockfish7 patches added if helpful)
- bug fixes and speed improvements
- ELO increase 100 ELO ( CCRL 40/4: 2750 for compiled VB6 EXE, 2250 for VBA code)


----------------------------------------------------------------
Changes in version 2.05 / 9-FEB-2016
 - fixed problem with Winboard.exe: removed: message box says that draws are not accepted

Changes in version 2.04 / 27-FEB-2016
 - fixed problem with Winboard.exe with black to move

Changes in version 2.03 
 - fixed WORD VBA problem with Avast virus scanner: creates false positive alert => AutoOpen removed fixed it

Changes in version 2.02
 - fixed VBA GUI problem with promotions

Changes in version 2.01
 - show endgame table base access in VBA GUI


Changes from version 1.02 to 2.0 (Roger Zuehlsdorf 14-FEB-2016)
----------------------------------------------------------------

- improved evalution and search

- problem fixes in hash logic and castling rights

- added special functions for endgames with KQKP, KQKR, KRKP

- added endgame table base access to web service for root position and first ply (see ini file)

- added Wb2Uci files for use with UCI interface

- ELO increase: 100 ELO 


Playing strength (Test matches: CCRL conditions 40  moves in 4 minutes, Powerbook2010-8 moves):
 

Development/Modules/EPD.bas
Attribute VB_Name = "EPDbas"
'==================================================
'= EPDBas:
'= EPD file format handling
'==================================================
Option Explicit
' Table for board indexes
Private EPDTable(63) As Long

'---------------------------------------------------------------------------
'ReadEPD()
'---------------------------------------------------------------------------
Public Function ReadEPD(ByVal sEpdString As String) As Boolean
  Dim NumSquares  As Long, i As Long
  Dim sChar       As String
  Dim arCmdList() As String, p As Long
  Erase arFiftyMove(): Erase GamePosHash(): Erase arGameMoves()
  Fifty = 0: GameMovesCnt = 0
  HintMove = EmptyMove
  arCmdList = Split(sEpdString)
  If UBound(arCmdList) < 3 Then
    ReadEPD = False
    Exit Function
  End If
  BookPly = BOOK_MAX_PLY + 1

  For i = 0 To 63
    Board(EPDTable(i)) = NO_PIECE
  Next

  For i = 0 To MAX_BOARD: Moved(0) = 1: Next ' set unknown moved status
  For i = 1 To Len(arCmdList(0))
    sChar = Mid$(arCmdList(0), i, 1)

    Select Case sChar
      Case "P"
        Board(EPDTable(NumSquares)) = WPAWN
        NumSquares = NumSquares + 1
      Case "p"
        Board(EPDTable(NumSquares)) = BPAWN
        NumSquares = NumSquares + 1
      Case "N"
        Board(EPDTable(NumSquares)) = WKNIGHT
        NumSquares = NumSquares + 1
      Case "n"
        Board(EPDTable(NumSquares)) = BKNIGHT
        NumSquares = NumSquares + 1
      Case "K"
        WKingLoc = EPDTable(NumSquares)
        Board(WKingLoc) = WKING
        NumSquares = NumSquares + 1
      Case "k"
        BKingLoc = EPDTable(NumSquares)
        Board(BKingLoc) = BKING
        NumSquares = NumSquares + 1
      Case "R"
        Board(EPDTable(NumSquares)) = WROOK
        NumSquares = NumSquares + 1
      Case "r"
        Board(EPDTable(NumSquares)) = BROOK
        NumSquares = NumSquares + 1
      Case "Q"
        Board(EPDTable(NumSquares)) = WQUEEN
        NumSquares = NumSquares + 1
      Case "q"
        Board(EPDTable(NumSquares)) = BQUEEN
        NumSquares = NumSquares + 1
      Case "B"
        Board(EPDTable(NumSquares)) = WBISHOP
        NumSquares = NumSquares + 1
      Case "b"
        Board(EPDTable(NumSquares)) = BBISHOP
        NumSquares = NumSquares + 1
      Case "/"
      Case Else
        NumSquares = NumSquares + Val(sChar)
    End Select

  Next

  ' part 2: color to move
  sChar = arCmdList(1)
  If LCase(sChar) = "w" Then
    bWhiteToMove = True
  ElseIf LCase(sChar) = "b" Then
    bWhiteToMove = False
  Else
    Exit Function
  End If
  bCompIsWhite = Not bWhiteToMove
  'Part 3: castling
  Moved(WKING_START) = 1: Moved(28) = 1: Moved(21) = 1
  Moved(BKING_START) = 1: Moved(98) = 1: Moved(91) = 1

  For i = 1 To Len(arCmdList(2))
    sChar = Mid$(arCmdList(2), i, 1)

    Select Case sChar
      Case "K"
        Moved(WKING_START) = 0
        Moved(28) = 0
      Case "Q"
        Moved(WKING_START) = 0
        Moved(21) = 0
      Case "k"
        Moved(BKING_START) = 0
        Moved(98) = 0
      Case "q"
        Moved(BKING_START) = 0
        Moved(91) = 0
      Case "-"
        Exit For
    End Select

  Next

  'Part4 : EnPassant
  sChar = arCmdList(3)
  If sChar <> "-" Then
    p = FileRev(Left$(sChar, 1)) + RankRev(Right$(sChar, 1))
    If bWhiteToMove Then
      If Right$(sChar, 1) = "6" Then
        Board(p) = BEP_PIECE: EpPosArr(1) = p
      End If
    Else
      If Right$(sChar, 1) = "3" Then
        Board(p) = WEP_PIECE: EpPosArr(1) = p
      End If
    End If
  End If
  'Part5 : Fifty move half move count
  If UBound(arCmdList) >= 4 Then
    sChar = arCmdList(4)
    If sChar <> "" Then
      If Val("0" & sChar) > 0 Then Fifty = Val(sChar)
    End If
  End If
  'Part5 : full move count: 1 before first move, 2 after first black move
  GameMovesCnt = 0
  If UBound(arCmdList) >= 5 Then
    sChar = arCmdList(5)
    If sChar <> "" Then
      If Val("0" & sChar) > 0 Then
        GameMovesCnt = GetMax(0, (Val(sChar) - 1) * 2)
        If Not bWhiteToMove Then GameMovesCnt = GameMovesCnt + 1
      End If
    End If
  End If
  InitPieceSquares
  GamePosHash(GameMovesCnt) = HashBoard(EmptyMove) ' for 3x repetition draw
  ReadEPD = True
End Function

'---------------------------------------------------------------------------
'WriteEPD() -
'---------------------------------------------------------------------------
Public Function WriteEPD() As String
  Dim i        As Long
  Dim iPiece   As Long, iEmptySquares As Long
  Dim sEPD     As String, sRow As String
  Dim sEPPiece As String, sCastle As String
  sEPPiece = "-"

  For i = 0 To 63
    If i Mod 8 = 0 And i > 0 Then
      sEPD = sEPD & "/" & sRow & Format$(iEmptySquares, "#")
      iEmptySquares = 0
      sRow = ""
    End If
    iPiece = Board(EPDTable(i))

    Select Case iPiece
      Case NO_PIECE
        iEmptySquares = iEmptySquares + 1
      Case WEP_PIECE, BEP_PIECE
        sEPPiece = Chr$(File(EPDTable(i)) + 96) & Rank(EPDTable(i))
        iEmptySquares = iEmptySquares + 1
      Case Else
        sRow = sRow & Format$(iEmptySquares, "#") & Piece2Alpha(iPiece)
        iEmptySquares = 0
    End Select

  Next

  sEPD = sEPD & "/" & sRow & Format$(iEmptySquares, "#")
  sEPD = Right$(sEPD, Len(sEPD) - 1)
  If bWhiteToMove Then
    sEPD = sEPD & " w"
  Else
    sEPD = sEPD & " b"
  End If
  If Moved(WKING_START) = 0 Then
    If Moved(28) = 0 Then sCastle = "K"
    If Moved(21) = 0 Then sCastle = sCastle & "Q"
  End If
  If Moved(BKING_START) = 0 Then
    If Moved(98) = 0 Then sCastle = sCastle & "k"
    If Moved(91) = 0 Then sCastle = sCastle & "q"
  End If
  If sCastle = "" Then sCastle = "-"
  sEPD = sEPD & " " & sCastle & " " & sEPPiece
  sEPD = sEPD & " " & CStr(Fifty)
  sEPD = sEPD & " " & CStr(GameMovesCnt \ 2 + 1)
  WriteEPD = sEPD
End Function

Public Sub InitEPDTable()
  EPDTable(0) = 91: EPDTable(1) = 92: EPDTable(2) = 93: EPDTable(3) = 94
  EPDTable(4) = 95: EPDTable(5) = 96: EPDTable(6) = 97: EPDTable(7) = 98
  EPDTable(8) = 81: EPDTable(9) = 82: EPDTable(10) = 83: EPDTable(11) = 84
  EPDTable(12) = 85: EPDTable(13) = 86: EPDTable(14) = 87: EPDTable(15) = 88
  EPDTable(16) = 71: EPDTable(17) = 72: EPDTable(18) = 73: EPDTable(19) = 74
  EPDTable(20) = 75: EPDTable(21) = 76: EPDTable(22) = 77: EPDTable(23) = 78
  EPDTable(24) = 61: EPDTable(25) = 62: EPDTable(26) = 63: EPDTable(27) = 64
  EPDTable(28) = 65: EPDTable(29) = 66: EPDTable(30) = 67: EPDTable(31) = 68
  EPDTable(32) = 51: EPDTable(33) = 52: EPDTable(34) = 53: EPDTable(35) = 54
  EPDTable(36) = 55: EPDTable(37) = 56: EPDTable(38) = 57: EPDTable(39) = 58
  EPDTable(40) = 41: EPDTable(41) = 42: EPDTable(42) = 43: EPDTable(43) = 44
  EPDTable(44) = 45: EPDTable(45) = 46: EPDTable(46) = 47: EPDTable(47) = 48
  EPDTable(48) = 31: EPDTable(49) = 32: EPDTable(50) = 33: EPDTable(51) = 34
  EPDTable(52) = 35: EPDTable(53) = 36: EPDTable(54) = 37: EPDTable(55) = 38
  EPDTable(56) = 21: EPDTable(57) = 22: EPDTable(58) = 23: EPDTable(59) = 24
  EPDTable(60) = 25: EPDTable(61) = 26: EPDTable(62) = 27: EPDTable(63) = 28
End Sub
